client/rpc.go

```diff
@@ -4,70 +4,254 @@ package client
+		return fmt.Errorf("sending json object: %w", err)
+	}
+	// Return error from async response cb.
+	select {
+	case err := <-errChan:
+		return err
+	case <-ctx.Done():
+		return ctx.Err()
+	}
 }
 
-func (r *RPC) GetDepositProof(ctx context.Context, epoch uint64, user common.Address) (reply tee.DepositProof, err error) {
-	for i := 0; ; i++ {
-		if e := r.call(ctx, "RemoteEnclave.GetDepositProof", user, &reply); e != nil {
-			return tee.DepositProof{}, e
+// Subscribe subscribes to the Balance and Deposit proof topic.
+// The user must always read the proofs via `DepositProof` and
+// `BalanceProof`. Calling this function more than once if it did not
+// error will cause undefined behaviour.
+func (r *RPC) Subscribe(ctx context.Context, user common.Address) (*Subscription, error) {
+	r.subscription = &Subscription{
+		Embedding: perunlog.MakeEmbedding(perunlog.WithField("role", "proofSub")),
+		// Buffer the proofs here, otherwise the client has to read them
+		// immediately to prevent that they get reordered by a race condition
+		// from the go-routines writing them to the channel since a mutex is
+		// not FIFO.
+		balProofs: make(chan tee.BalanceProof, 10),
+		depProofs: make(chan tee.DepositProof, 10),
+	}
+
+	call := wire.NewSubscribe(r.nextID(), user)
+	errChan := make(chan error)
+	// Setup async response cb.
+	r.registerCallback(call.Call.ID, func(result wire.Result, msg []byte) {
+		if result.Error != "" {
+			errChan <- fmt.Errorf("Subscribe RPC result: %s", result.Error)
+		} else {
+			errChan <- nil
+		}
+	})
+	// Make the call.
+	if err := r.sendJSON(call); err != nil {
+		return nil, fmt.Errorf("sending json object: %w", err)
+	}
+	// Return error from async response cb.
+	select {
+	case err := <-errChan:
+		return r.subscription, err
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	}
+}
+
+func (r *RPC) handleConnections() {
+	for !r.IsClosed() {
+		// gorilla has no async read method?!
+		_, data, err := r.conn.ReadMessage()
+		if err != nil {
+			r.Log().Error("reading ws message: ", err)
+			time.Sleep(1 * time.Second)
+			continue
 		}
-		if reply.Balance.Epoch < epoch {
-			time.Sleep(time.Second * 1)
+		r.Log().Trace("client received: ", string(data))
+		var msg wire.Result
+		if err := json.Unmarshal(data, &msg); err != nil {
+			r.Log().Error("decoding message: ", err)
 			continue
-		} else if reply.Balance.Epoch > epoch {
-			return tee.DepositProof{}, errors.New("Newer epoch")
 		}
+
+		switch {
+		case msg.ID != "":
+			r.callCallback(msg, data)
+		case msg.Topic != "":
+			if r.subscription == nil {
+				r.Log().Error("Received proof without subscription")
+				return
+			}
+			r.subscription.handleTopic(msg.Topic, data)
+		default:
+			r.Log().Error("Received result without ID or Topic")
+		}
+	}
+}
+
+func (r *RPC) registerCallback(id wire.ID, cb callback) {
+	r.cbMtx.Lock()
+	defer r.cbMtx.Unlock()
+	if _, ok := r.callbacks[id]; ok {
+		r.Log().Error("Callbacks should never be overridden.")
```

Comment:
"overwritten"

Since you're writing an error to the log here and not return an error, this should be rephrased to a sentence like
```suggestion
		r.Log().WithField("id", id).Error("Callback already registered, skipping.")
```